### PR TITLE
Adding first version of MOOSE SQA docs

### LINKED ISSUE #6335: Import SQA Documentation into the MOOSE repository
This ticket shall be referenced for importing all of the MOOSE related SQA documents into the framework.

### PR DESCRIPTION
**Pull request by permcody**
Contains the following QA documents:
- MOOSE Software Requirements Specification (SRS)
- MOOSE Software Design Document (SDD)
- MOOSE Requirements Tracability Matrix (auto-generated)

closes #6335 

@idaholab/moose-developers - This is it, the first set of actual documents that gets us closer to NQA-1 practices and makes our development process more transparent. None of this is set in stone but we are getting pressure to make this happen very soon.

For the review you are welcome to pick over the contents of the TeX file all you want. I just need a place to get started. This won't really impact our process much but as new systems go in, we'll certainly need to modify and update this document (which is why it'll live in the repo). The bigger change will be with how the requirements are tested and traced. I'm planning to annotate several of our basic systems tests so that the `tracability_matrix.py`script will pick them up and fill out the table we need for one of our check marks. The syntax I choose for that is simply `@Requirement <some alphanumeric designation>` If you hate it, let me know. If you love it, let me know that too.

This is important, let's get your feedback.

### CODE REVIEW

**Comment by permcody** [comment]
@WilkAndy now that I'm not on my phone, I'll address your comments. First thanks for taking a look at this. We are trying to qualify our software under a nuclear quality assurance program that is required for processes and software used in a "safety" roll in nuclear applications. We aren't anywhere close to deploying but someday we might and then these more stringent processes will be heavily scrutinized.

Now to address your comments:
 1. Thanks, I'll see if I can break the large paragraph
 2. 24 -> 48 hours (that's a good comment and may be fine, I'll have to check the reqs)
 3. I'd like to 80% on modules eventually if not today. We definitely have that requirement in the framework so it's only natural to make it a requirement in the modules too.
 4. When we reach 80% in the modules, we'll certainly make CIVET fail when that threshold isn't reached. We used to do this in the framework but that requirement got dropped somewhere along the way. @brianmoose - let's add this back in if we can, we should be able to inspect the coverage report for that percentage.
 5. Oops, that's a typo
 6. I'd like to avoid committing the generated PDF. I think a better idea would be to have a way to make our webpage render it on demand. That would be cool and could be useful for documentation as well.
 7. I'm glad you agree, and we do rely on it for our quality but it doesn't do everything we need it to do for what a normal Quality Assurance program needs. They need even more assurances. We should probably take more credit than we currently do though, you're right.
 
Thanks for looking this over. There will be more effort in this area in the future as we continue down this path.

**Comment by brianmoose** [comment]
The check for 80% coverage on the framework is already in place. On the master "Infrastructure" recipe in the step "Moose code coverage". It will fail if it drops below 80%.

**Comment by andrsd** [comment]
> I'd like to 80% on modules eventually if not today. We definitely have that requirement in the framework so it's only natural to make it a requirement in the modules too.

One thing to think about is, do we want 80% for all modules together or guard it on per module basis...

**Comment by permcody** [comment]
Per module basis, otherwise some people will get away with less than the
minimum because they are riding on better tested modules. I was thinking
about that too.

**Comment by WilkAndy** [comment]
There used to be a webpage or something where i could see which lines of code were tested.  Does this still exist?  If so, where?

**Comment by friedmud** [comment]
http://mooseframework.org/docs/coverage/moose/coverage/

You can get to it from the "Test Coverage" section of the Documentation
page here:

http://mooseframework.org/documentation/

Derek

**Comment by WilkAndy** [comment]
Thanks @friedmud .   What about modules coverage?

**Comment by friedmud** [comment]
I don't know if it's done currently... @permcody would know...

**Comment by permcody** [comment]
We're not, but we're about to add it. Time to see which of our modules are
well tested and which are not. Stay tuned!

**Comment by WilkAndy** [comment]
Regarding the 80% - for me it would be most useful if it were on a per-class level.  Then when i make a new class i can be told "sorry, checks failed due to less than 80% test coverage" and i can immediately fix that.

**Comment by permcody** [comment]
That's an interesting idea and actually somewhat useful as our code base continues to grow. I'll think about.

**Comment by aeslaughter** [comment]
I think I like this idea as well.

**Comment by brianmoose** [comment]
I added the generation of code coverage for each module. There are links on mooseframework.org/documentation.
For the lazy, here is the current status:
chemical_reactions: 20.9%
combined: 94.7%
contact: 88.1%
heat_conduction: 88.3%
linear_elasticity: 85.0%
misc: 47.2%
navier_stokes: 14.1%
phase_field: 88.0%
richards: 91.1%
solid_mechanics: 79.2%
tensor_mechanics: 85.7%
water_steam_eos: 91.4%

**Comment by permcody** [comment]
Holy Dooley!
Chemical Reactions => FAIL
Navier Stokes => FAIL

Richards => 90+% Nicely Done Andy

For those of you who have worked with these modules, it would be really
nice to get some better coverage in there.
Cody

**Comment by jwpeterson** [comment]
On Wed, Feb 24, 2016 at 8:44 AM, Cody Permann <notifications@github.com>
wrote:

> Holy Dooley!
> Chemical Reactions => FAIL
> Navier Stokes => FAIL
>

> navier_stokes: 14.1%

I should be able to improve this number significantly by adding just a 1 or
2 tests... let me take a look.

-- 
John

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+threading and compiling while providing an easy to use interface targeted at engineers that may not have a lot of
+software development experience. MOOSE will require extreme scalability and flexibility when compared to other FEM
+frameworks. For instance, MOOSE needs the ability to run extremely complex material models, or even third-party applications
+within a parallel simulation without sacrificing parallelism. This capability is in contrast to what is often seen in
+commercial packages, where custom material models can limit the parallel scalability, forcing serial runs in the most severe
+cases. When comparing high-end capabilities, many MOOSE competitors target modest-sized clusters with just a few thousand
+processing cores. MOOSE, however, will be required to routinely executed on much larger clusters with scalability to clusters
+available in the top 500 systems. (top500.org). MOOSE will also be targeted at smaller systems such as high-end laptop
+computers.
+
```
Comment (aeslaughter) [inline framework/doc/sqa/SoftwareRequirements.tex]: We should probably do this then, don't we only have a study up to 10,000

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+    \rowcolor{gray}
+    Number & Requirement Description \\
+    F1.10 & The system shall allow support for user-defined object for controlling the execution stages of a simulation. \\ \hline
+    F1.20 & The system shall allow for user-defined time step selection for simulations using a time-based execution scheme. \\ \hline
+    F1.30 & The system shall allow for user-defined time step integration schemes for simulations using a time-based execution scheme. \\ \hline
+    F1.40 & The system shall support user-defined matrix preconditioning that may be applied during the solve stages. \\ \hline
+    F1.50 & The system should support a programatic method for building up the necessary objects necessary for a simulation. \\ \hline
+    F1.60 & The system shall provide the ability to resume a previous simulation due to fault or intentional termination. \\ \hline
+    F1.70 & The system shall allow for user-defined output types for simulation data. \\ \hline
+    F1.80 & Predictors \\ \hline
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: "The system shall provide a method of providing improved initial guesses (also known as "predictions") for the solution at subsequent timesteps."

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+on any POSIX compliant system. The software will also make use FEM and numerical libraries that run on POSIX
+systems as well. The main interface for the software will be command line based with no assumptions requiing advanced
+terminal capabilities such as coloring and line control.
+
+\section{References}
+
+\begin{itemize}
+\item ASME NQA 1 2008 with the NQA-1a-2009 addenda, “Quality Assurance Requirements for Nuclear Facility Applications,” First Edition, August 31, 2009.
+\item ISO/IEC/IEEE 24765:2010(E), “Systems and software engineering — Vocabulary,” First Edition, December 15, 2010.
+\item LWP 13620, “Managing Information Technology Assets,” Rev. ?, ?.
```
Comment (aeslaughter) [inline framework/doc/sqa/SoftwareRequirements.tex]: ??

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+\begin{table}[!htbp]
+  \caption{Minimum Requirements.\label{tab:req}}
+  \begin{tabular}{|p{12cm}|}
+    \rowcolor{gray}
+    Minimum Requirements \\
+    \begin{itemize}
+    \item A POSIX compliant Unix including the two most recent versions of OS X and most current versions of Linux.
+    \item RAM: 4 GB for optimized compilation (8 GB for debug compilation), 2 GB per core execution
+    \item Disk: 100 GB
+    \item Compilers: GCC, Clang, and Intel
```
Comment (aeslaughter) [inline framework/doc/sqa/SoftwareRequirements.tex]: or Intel

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+    \rowcolor{gray}
+    Number & Requirement Description \\ \hline
+    F4.10 & The system shall allow for the creation of objects that represent physical material properties based on physical quantities. \\ \hline
+  \end{tabular}
+\end{table}
+
+
+\begin{table}[!htbp]
+  \caption{Function Name: Auxilary Calculations.\label{tab:aux}}
+  \newcounter{MyFifthTableCounter}
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: Delete these MyTableCounter things if you are no longer using them.

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+This document provides design details pertinent to application developers as well as framework developers.
+
+\subsection{Use Cases}
+The MOOSE Framework is targeted at two main groups of actors: Developers and Users. Developers are the main
+use case. These are typically students and professionals trained in science and engineering fields with some
+level of experience with coding but typically very little formal software development training. The other user
+group is Users. Those who intend to use an application built upon the framework without writing any computer
+code themselves. Instead they may modify or create input files for driving a simulation, run the application, and
+analyze the results. All interactions through MOOSE are primarily through the command-line interface and through
+a customizable block-based input file described later (REFERENCE NEEDED).
```
Comment (aeslaughter) [inline framework/doc/sqa/SoftwareRequirements.tex]: Ref. missing

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+\item Flaws identified in the system shall be reported and tracked in a ticket or issue based system. The technical lead
+      will determine the severity and priority of all reported issues and assign resources at his or her discretion to
+      resolve identified issues.
+\item The software maintainers will entertain all proposed changes to the system in a timely manner (within two business days).
+\item The core framework in its entirity will be made publically available under the LGPL version 2.0 license.
+\end{enumerate}
+
+\subsubsection{Reliability}
+The regression test suite will cover at least 80\% of all lines of code at all times. Known regressions will be recorded
+and tracked (see \ref{Maintainability}) through to an independent and satisfactory resolution.
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: Delete "through"

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+
+\subsection{Information Management}
+The core framework in its entirety will be made publically available on an appropriate repository hosting site. Backups and
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: ~~publically~~ publicly

File: framework/doc/sqa/SoftwareRequirements.tex (hunk 1 above)
Comment (permcody) [inline framework/doc/sqa/SoftwareRequirements.tex]: Agreed, This is part of a general overview of the the project. I don't have any requirements tied to scalability at this point. We may add that in the future if we want to architect a way to test this requirement.

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+analyze the results. All interactions through MOOSE are primarily through the command-line interface and through
+a customizable block-based input file described later (REFERENCE NEEDED).
+
+\subsection{Software Overview}
+
+\subsubsection{Introduction}
+The MOOSE framework itself is composed of a wide range of plugable systems. Each system is generally composed of a single or small set of C++ objects intended to be specialized by a Developer to solve a specific problem. To accomplish this design goal, MOOSE uses several modern object-oriented design patterns. The primary overarching pattern is the ``Factory Pattern''. User's needing to extend MOOSE may inherit from one of the public interfaces overridden the designated methods to accomplish a specific task. The design of each of these systems is documented on the mooseframework.org wiki in the Tutorial section. Additionally, up-to-date documentation extracted from the source is maintained on the the mooseframework.org documentation site after every successful merge to MOOSE's stable branch. After these objects are created, the can be registered with the framework and used immediately in a MOOSE input file.
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: ~~plugable~~ pluggable

Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: This is all one huge line please split it up.

Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: ~~User's needing~~ Users needing

Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: This sentence makes no sense.
"may inherit from one of the public interfaces overridden the designated methods to accomplish a specific task"

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+
+\subsubsection{System Architecture}
+The MOOSE framework architecture consists of a core and several plugable systems. The core of MOOSE consists of a number of key objects responsible for setting up and managing the user-defined objects of a finite element simulation. This core set of objects has limited extendability and exist for every simulation configuration that the framework is capable of running.
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: ~~plugable~~ pluggable

File: framework/doc/sqa/README [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+the documents related to the MOOSE framework software project:
+
+Software Requirements Specification
+Software Design Document
+  To generate run make
+  requires pdflatex and make
+
+Requirements Tracability Matrix
+  To generate run ./tracability_matrix.py
+  requires Python 2.7 and ReportLab (installable through conda)
```
Comment (jwpeterson) [inline framework/doc/sqa/README]: Need newline at EOF.

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+several solution strategies. Small perturbations in strongly-coupled parameters often have very large adverse effects on
+convergence behavior. These adverse effects are compounded as additional physics are added to a model. To overcome these
+challenges, MOOSE employs three distinct yet compatible systems for solving these types of problems.
+
+First, an advanced numerical technique called the Jacobian-Free Newton-Krylov (JFNK) method is employed to solve the most fully-coupled physics
+in an accurate, consistent way. An example of this would be the effect of temperature on the expansion or contraction of a
+material. While the JFNK numerical method is very effective at solving fully-coupled equations, it can also be computationally
+expensive. Plus, not all physical phenomena in a given model are truly coupled to one another. For instance, in a reactor, the
+speed of the coolant flow may not have any direct effect on the complex chemical reactions taking place inside the fuel rods.
+We call such models “loosely-coupled.” A robust, scalable system must strike the proper balance between the various modeling
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: There are non-ASCII quote characters around "loosely-coupled"

File: framework/doc/sqa/SoftwareRequirements.tex (hunk 3 above)
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: Fix these question marks...

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+The software should be designed with the fewest possible constraints. Ideally the software should run on a wide
+variety of evolving hardware so it should follow well-adopted standards and guidlines. The software should run
+on any POSIX compliant system. The software will also make use FEM and numerical libraries that run on POSIX
+systems as well. The main interface for the software will be command line based with no assumptions requiing advanced
+terminal capabilities such as coloring and line control.
+
+\section{References}
+
+\begin{itemize}
+\item ASME NQA 1 2008 with the NQA-1a-2009 addenda, “Quality Assurance Requirements for Nuclear Facility Applications,” First Edition, August 31, 2009.
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: Check the quotes on this line.

File: framework/doc/sqa/SoftwareRequirements.tex [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+\item ISO/IEC/IEEE 24765:2010(E), “Systems and software engineering — Vocabulary,” First Edition, December 15, 2010.
```
Comment (jwpeterson) [inline framework/doc/sqa/SoftwareRequirements.tex]: Check quotes

**Pull request merged**